x-pack/plugins/apm/typings/Service.ts

```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export interface IServiceListItem {
+  serviceName: string;
+  agentName: string | undefined;
+  transactionsPerMinute: number;
+  errorsPerMinute: number;
+  avgResponseTime: number;
+}
```

Comment:
A follow up question I would have is: do we want to have many different "service" types or should we try to align them, perhaps along with some kind of `Partial` implementation?

Reply:
> My understanding of "why does something go in top-level typings" was if it's shared between server and client, which seems to make sense to me. Thoughts?

I think interfaces should live as close to the source as possible. I don't think it changes anything that the interface is consumed by the client, more to the contrary. Being able to link the client to the backend can actually be a huge advantage.

The rest methods is our link between the frontend and the backend but it's super hard as a dev to figure out this link. Eg. what backend method is this frontend method calling?
https://github.com/sqren/kibana/blob/52036194b4ced069d43f1d2d924a64a9755159a2/x-pack/plugins/apm/public/services/rest/ml.js#L11

With a type defined at source in the backend, it becomes super easy to find the backend method:
https://github.com/sqren/kibana/blob/52036194b4ced069d43f1d2d924a64a9755159a2/x-pack/plugins/apm/public/services/rest/apm.ts#L72-L86

Reply:
> do we want to have many different "service" types or should we try to align them, perhaps along with some kind of Partial implementation?

If aligning makes sense, I think we should do it. But here I'm mostly thinking about using the same names for the same props. 
I don't think we should send a lot of unnecessary data to the client, just to align two endpoints.

Service list item:
```
{
  service_name: string;
  agent_name: string | undefined;
  transactions_per_minute: number;
  errors_per_minute: number;
  avg_response_time: number;
}

Service:
```ts
{
  service_name: string;
  types: string[];
  agent_name?: string;
}
```

Reply:
I like all of this in theory, but I'm not sure I understand the point of the typings folder as-is. I think in my envisioned perfect world, we would have a type for each of the main concepts in our system (Transaction, Span, Service, Trace, etc), with the variations just being pick/omit/extend versions of each but appearing in a single file for readability. Then the API response interfaces might be defined in the files that need them, and compose together the "master types" as necessary. 

```
// typings/Service.ts
interface ApmService {
  name: string;
  agentName?: string;
  transactionsPerMinute?: number;
  errorsPerMinute?: number;
  avgResponseTime?: number;
  types?: string[];
}

// possibly define list vs single versions here also

// get_services.ts
export type ServiceListResponse = ApmService[] // perhaps with Omit<>

// get_service.ts
export type ServiceResponse = ApmService // perhaps with Omit<>
```

My question about alignment was me wondering why the list and the single return different keys for the services and if that will ever get us into trouble with assuming we have a "service" object but it's one way in the list and another way in the single entry (especially with the single entry containing _less keys_ than the list item). If they were aligned, we wouldn't need the `Omit<>` at all (but maybe that's not necessary).

> I don't think we should send unnecessary data to the client, just to align two endpoints.

Makes sense, but I also think it makes sense to just always return a "service" when the client asks for a service (with the one exception being maybe a list endpoint only sends a smaller "summary" rather than including the entire resource for each item in what could be a long list), so I'm not sure where I land on that.

For now, since these aren't aligned, I'll move this back to the file and rename it. We can continue to bike-shed on the typescript stuff in another ticket. :)

Reply:
> I'm not sure I understand the point of the typings folder as-is

I agree. Right now it's a mess. We should look into that.

> I think in my envisioned perfect world, we would have a type for each of the main concepts in our system (Transaction, Span, Service, Trace, etc), with the variations just being pick/omit/extend versions of each but appearing in a single file for readability. Then the API response interfaces might be defined in the files that need them, and compose together the "master types" as necessary.

I like this idea if we can make it work with the constrains (that API endpoints are not always aligned).

> If they were aligned, we wouldn't need the Omit<> at all (but maybe that's not necessary).

We could make them 99% aligned if we added `transactions_per_minute`, `errors_per_minute` and `avg_response_time` to the `get_service` endpoint. That would require us do do the following extra aggs:
```js
{
  aggs: {
    avg: {
      avg: { field: TRANSACTION_DURATION }
    },
    events: {
      terms: { field: PROCESSOR_EVENT, size: 2 }
    }
  }
};
```
It's not a lot but it feels a little weird to add more code, so the client will fetch more data, that we are not going to show.
But if the consistency can aid in the mental models of the various data structures it might be worth it.

Reply:
Yeah ... I mean it's a good point you make. I honestly could argue it either way. I'll leave it alone for now and we'll keep thinking on it!